src/dbnode/storage/index/aggregate_results.go

```diff
@@ -0,0 +1,237 @@
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package index
+
+import (
+	"bytes"
+	"fmt"
+	"sync"
+
+	"github.com/m3db/m3/src/m3ninx/doc"
+	"github.com/m3db/m3x/ident"
+	"github.com/m3db/m3x/pool"
+)
+
+const missingDocumentFields = "invalid document fields: empty %s"
+
+type aggregatedResults struct {
+	sync.RWMutex
+
+	nsID          ident.ID
+	aggregateOpts AggregateResultsOptions
+
+	resultsMap *AggregateResultsMap
+
+	idPool    ident.Pool
+	bytesPool pool.CheckedBytesPool
+
+	pool       AggregateResultsPool
+	valuesPool AggregateValuesPool
+}
+
+// NewAggregateResults returns a new AggregateResults object.
+func NewAggregateResults(
+	namespaceID ident.ID,
+	aggregateOpts AggregateResultsOptions,
+	opts Options,
+) AggregateResults {
+	return &aggregatedResults{
+		nsID:          namespaceID,
+		aggregateOpts: aggregateOpts,
+		resultsMap:    newAggregateResultsMap(opts.IdentifierPool()),
+		idPool:        opts.IdentifierPool(),
+		bytesPool:     opts.CheckedBytesPool(),
+		pool:          opts.AggregateResultsPool(),
+		valuesPool:    opts.AggregateValuesPool(),
+	}
+}
+
+func (r *aggregatedResults) Reset(
+	nsID ident.ID,
+	aggregateOpts AggregateResultsOptions,
+) {
+	r.Lock()
+
+	r.aggregateOpts = aggregateOpts
+
+	// finalize existing held nsID
+	if r.nsID != nil {
+		r.nsID.Finalize()
+	}
+
+	// make an independent copy of the new nsID
+	if nsID != nil {
+		nsID = r.idPool.Clone(nsID)
+	}
+	r.nsID = nsID
+
+	// reset all values from map first
+	for _, entry := range r.resultsMap.Iter() {
+		valueMap := entry.Value()
+		valueMap.reset()
+	}
+
+	// reset all keys in the map next
+	r.resultsMap.Reset()
+
+	// NB: could do keys+value in one step but I'm trying to avoid
+	// using an internal method of a code-gen'd type.
+	r.Unlock()
+}
+
+func (r *aggregatedResults) AddDocuments(batch []doc.Document) (int, error) {
+	r.Lock()
+	err := r.addDocumentsBatchWithLock(batch)
+	size := r.resultsMap.Len()
+	r.Unlock()
+	return size, err
+}
+
+func (r *aggregatedResults) addDocumentsBatchWithLock(
+	batch []doc.Document,
+) error {
+	for i := range batch {
+		err := r.addDocumentWithLock(batch[i])
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *aggregatedResults) addDocumentWithLock(
+	document doc.Document,
+) error {
+	for _, field := range document.Fields {
+		if err := r.addFieldWithLock(field.Name, field.Value); err != nil {
+			return err
```

Comment:
nit: would be useful to bundle with document here too, something like: 
```return fmt.Errorf("unable to add document [%+v]: %v", document, err)```

Reply:
SGTM